From 98d5eaedd4ee8d36ec13bd38684fa454971d9cfb Mon Sep 17 00:00:00 2001
From: "Chen, Yu" <yu.y.chen@intel.com>
Date: Tue, 11 Jun 2019 14:13:18 +0800
Subject: [PATCH 1/3] Fix media scanner crash

Media scanner crash due to Device access poinit /product
is a link, switch to getCanonicalFile for the security check.

Change-Id: Icd44f8922288e41d5c74b768dbd985211cdc582a
Tracked-On:
Signed-off-by: Chen, Yu <yu.y.chen@intel.com>
Signed-off-by: Wang, ArvinX <arvinx.wang@intel.com>
---
 src/com/android/providers/media/MediaProvider.java | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/com/android/providers/media/MediaProvider.java b/src/com/android/providers/media/MediaProvider.java
index bbfc212c..39420757 100644
--- a/src/com/android/providers/media/MediaProvider.java
+++ b/src/com/android/providers/media/MediaProvider.java
@@ -2176,7 +2176,11 @@ public class MediaProvider extends ContentProvider {
         try {
             // Sanity check that the requested path actually lives on volume
             final String volumeName = resolveVolumeName(uri);
-            final Collection<File> allowed = getVolumeScanPaths(volumeName);
+            // Maybe the VolumeScanPath is a symbolic link, get the factual address.
+            final Collection<File> allowed = new ArrayList<>();
+            for (File file : getVolumeScanPaths(volumeName)) {
+                allowed.add(file.getCanonicalFile());
+            }
             final File actual = new File(values.getAsString(MediaColumns.DATA))
                     .getCanonicalFile();
             if (!FileUtils.contains(allowed, actual)) {
--
2.21.0
